// telegram/events/common.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventCommon = exports.EventBuilder = exports._intoIdSet = void 0;
const tl_1 = require("../tl");
const custom_1 = require("../tl/custom");
const Helpers_1 = require("../Helpers");
const __1 = require("../");
/** @hidden */
function _intoIdSet(client, chats) {
    return __awaiter(this, void 0, void 0, function* () {
        if (chats == undefined) {
            return undefined;
        }
        if (!Helpers_1.isArrayLike(chats)) {
            chats = [chats];
        }
        const result = new Set();
        for (let chat of chats) {
            if (typeof chat == "number") {
                if (chat < 0) {
                    result.add(chat);
                }
                else {
                    result.add(__1.utils.getPeerId(new tl_1.Api.PeerUser({
                        userId: chat,
                    })));
                    result.add(__1.utils.getPeerId(new tl_1.Api.PeerChat({
                        chatId: chat,
                    })));
                    result.add(__1.utils.getPeerId(new tl_1.Api.PeerChannel({
                        channelId: chat,
                    })));
                }
            }
            else if (typeof chat == "object" &&
                chat.SUBCLASS_OF_ID == 0x2d45687) {
                result.add(__1.utils.getPeerId(chat));
            }
            else {
                chat = yield client.getInputEntity(chat);
                if (chat instanceof tl_1.Api.InputPeerSelf) {
                    chat = yield client.getMe(true);
                }
                result.add(__1.utils.getPeerId(chat));
            }
        }
        return Array.from(result);
    });
}
exports._intoIdSet = _intoIdSet;
/**
 * The common event builder, with builtin support to filter per chat.<br/>
 * All events inherit this.
 */
class EventBuilder {
    constructor(eventParams) {
        this.chats = eventParams.chats;
        this.blacklistChats = eventParams.blacklistChats || false;
        this.resolved = false;
        this.func = eventParams.func;
    }
    build(update, others, callback) {
        if (update)
            return update;
    }
    resolve(client) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.resolved) {
                return;
            }
            yield this._resolve(client);
            this.resolved = true;
        });
    }
    _resolve(client) {
        return __awaiter(this, void 0, void 0, function* () {
            this.chats = yield _intoIdSet(client, this.chats);
        });
    }
    filter(event) {
        if (!this.resolved) {
            return;
        }
        if (this.chats != undefined && event.chatId != undefined) {
            const inside = this.chats.includes(event.chatId);
            if (inside == this.blacklistChats) {
                // If this chat matches but it's a blacklist ignore.
                // If it doesn't match but it's a whitelist ignore.
                return;
            }
        }
        if (this.func && !this.func(event)) {
            return;
        }
        return event;
    }
}
exports.EventBuilder = EventBuilder;
class EventCommon extends custom_1.ChatGetter {
    constructor({ chatPeer = undefined, msgId = undefined, broadcast = undefined, }) {
        super({ chatPeer, broadcast });
        this._eventName = "Event";
        this._entities = new Map();
        this._client = undefined;
        this._messageId = msgId;
    }
    _setClient(client) {
        this._client = client;
    }
    get client() {
        return this._client;
    }
}
exports.EventCommon = EventCommon;
