// telegram/tl/custom/chatGetter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatGetter = void 0;
const __1 = require("../../");
const api_1 = require("../api");
const util_1 = require("util");
const Helpers_1 = require("../../Helpers");
class ChatGetter {
    constructor({ chatPeer, inputChat, chat, broadcast, }) {
        ChatGetter.initClass(this, { chatPeer, inputChat, chat, broadcast });
    }
    [util_1.inspect.custom]() {
        return Helpers_1.betterConsoleLog(this);
    }
    static initClass(c, { chatPeer, inputChat, chat, broadcast }) {
        c._chatPeer = chatPeer;
        c._inputChat = inputChat;
        c._chat = chat;
        c._broadcast = broadcast;
        c._client = undefined;
    }
    get chat() {
        return this._chat;
    }
    getChat() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (!this._chat ||
                ("min" in this._chat && (yield this.getInputChat()))) {
                try {
                    if (this._inputChat) {
                        this._chat = yield ((_a = this._client) === null || _a === void 0 ? void 0 : _a.getEntity(this._inputChat));
                    }
                }
                catch (e) {
                    yield this._refetchChat();
                }
            }
            return this._chat;
        });
    }
    get inputChat() {
        if (!this._inputChat && this._chatPeer && this._client) {
            try {
                this._inputChat = this._client._entityCache.get(this._chatPeer);
            }
            catch (e) { }
        }
        return this._inputChat;
    }
    getInputChat() {
        return __awaiter(this, void 0, void 0, function* () {
            /*
            if (!this._inputChat && this.chatId && this._client) {
                try {
                    const target = this.chatId;
                    for await (const dialog of this._client.iterDialogs(100)) {
                        if (dialog.id === target) {
                            this._chat = dialog.entity;
                            this._inputChat = dialog.inputEntity;
                            break;
                        }
                    }
                } catch (e) {
                    // do nothing
                }
                return this._inputChat;
            }*/
            return this._inputChat;
        });
    }
    get chatId() {
        return this._chatPeer ? __1.utils.getPeerId(this._chatPeer) : undefined;
    }
    get isPrivate() {
        return this._chatPeer
            ? this._chatPeer instanceof api_1.Api.PeerUser
            : undefined;
    }
    get isGroup() {
        if (!this._broadcast && this.chat && "broadcast" in this.chat) {
            this._broadcast = Boolean(this.chat.broadcast);
        }
        if (this._chatPeer instanceof api_1.Api.PeerChannel) {
            if (this._broadcast === undefined) {
                return undefined;
            }
            else {
                return !this._broadcast;
            }
        }
        return this._chatPeer instanceof api_1.Api.PeerChat;
    }
    get isChannel() {
        return this._chatPeer instanceof api_1.Api.PeerChannel;
    }
    _refetchChat() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
}
exports.ChatGetter = ChatGetter;
