// telegram/tl/custom/senderGetter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SenderGetter = void 0;
const api_1 = require("../api");
const util_1 = require("util");
const Helpers_1 = require("../../Helpers");
class SenderGetter {
    constructor({ senderId, sender, inputSender, }) {
        SenderGetter.initClass(this, { senderId, sender, inputSender });
    }
    [util_1.inspect.custom]() {
        return Helpers_1.betterConsoleLog(this);
    }
    static initClass(c, { senderId, sender, inputSender }) {
        c._senderId = senderId;
        c._sender = sender;
        c._inputSender = inputSender;
        c._client = undefined;
    }
    get sender() {
        return this._sender;
    }
    getSender() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this._client &&
                (!this._sender ||
                    (this._sender instanceof api_1.Api.Channel && this._sender.min)) &&
                (yield this.getInputSender())) {
                try {
                    this._sender = yield this._client.getEntity(this._inputSender);
                }
                catch (e) {
                    yield this._refetchSender();
                }
            }
            return this._sender;
        });
    }
    get inputSender() {
        if (!this._inputSender && this._senderId && this._client) {
            try {
                this._inputSender = this._client._entityCache.get(this._senderId);
            }
            catch (e) { }
        }
        return this._inputSender;
    }
    getInputSender() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.inputSender && this._senderId && this._client) {
                yield this._refetchSender();
            }
            return this._inputSender;
        });
    }
    get senderId() {
        return this._senderId;
    }
    _refetchSender() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
}
exports.SenderGetter = SenderGetter;
