// telegram/client/updates.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports._updateLoop = exports._dispatchUpdate = exports._processUpdate = exports._handleUpdate = exports.catchUp = exports.listEventHandlers = exports.removeEventHandler = exports.addEventHandler = exports.on = void 0;
const tl_1 = require("../tl");
const __1 = require("../");
const big_integer_1 = __importDefault(require("big-integer"));
const network_1 = require("../network");
const index_1 = require("../index");
// export class UpdateMethods
function on(client, event) {
    return (f) => {
        client.addEventHandler(f, event);
        return f;
    };
}
exports.on = on;
function addEventHandler(client, callback, event) {
    if (event == undefined) {
        // recursive imports :(
        const raw = require("../events/Raw").Raw;
        event = new raw({});
    }
    event.client = client;
    client._eventBuilders.push([event, callback]);
}
exports.addEventHandler = addEventHandler;
function removeEventHandler(client, callback, event) {
    client._eventBuilders = client._eventBuilders.filter(function (item) {
        return item !== [event, callback];
    });
}
exports.removeEventHandler = removeEventHandler;
function listEventHandlers(client) {
    return client._eventBuilders;
}
exports.listEventHandlers = listEventHandlers;
function catchUp() {
    // TODO
}
exports.catchUp = catchUp;
function _handleUpdate(client, update) {
    if (typeof update === "number") {
        if ([-1, 0, 1].includes(update)) {
            _dispatchUpdate(client, {
                update: new network_1.UpdateConnectionState(update),
            });
            return;
        }
    }
    //this.session.processEntities(update)
    client._entityCache.add(update);
    client.session.processEntities(update);
    if (update instanceof tl_1.Api.Updates ||
        update instanceof tl_1.Api.UpdatesCombined) {
        // TODO deal with entities
        const entities = new Map();
        for (const x of [...update.users, ...update.chats]) {
            entities.set(index_1.utils.getPeerId(x), x);
        }
        for (const u of update.updates) {
            _processUpdate(client, u, update.updates, entities);
        }
    }
    else if (update instanceof tl_1.Api.UpdateShort) {
        _processUpdate(client, update.update, null);
    }
    else {
        _processUpdate(client, update, null);
    }
}
exports._handleUpdate = _handleUpdate;
function _processUpdate(client, update, others, entities) {
    update._entities = entities || new Map();
    const args = {
        update: update,
        others: others,
    };
    _dispatchUpdate(client, args);
}
exports._processUpdate = _processUpdate;
function _dispatchUpdate(client, args) {
    return __awaiter(this, void 0, void 0, function* () {
        for (const [builder, callback] of client._eventBuilders) {
            if (!builder.resolved) {
                yield builder.resolve(client);
            }
            let event = args.update;
            if (event) {
                if (!client._selfInputPeer) {
                    yield client.getMe(true);
                }
                if (!(event instanceof network_1.UpdateConnectionState)) {
                    // TODO fix me
                }
                // TODO fix others not being passed
                event = builder.build(event, undefined, callback);
                if (event) {
                    if ("_eventName" in event) {
                        event._setClient(client);
                        event.originalUpdate = args.update;
                        event._entities = args.update._entities;
                    }
                    const filter = yield builder.filter(event);
                    if (!filter) {
                        continue;
                    }
                    try {
                        yield callback(event);
                    }
                    catch (e) {
                        console.error(e);
                    }
                }
            }
        }
    });
}
exports._dispatchUpdate = _dispatchUpdate;
function _updateLoop(client) {
    return __awaiter(this, void 0, void 0, function* () {
        while (client.connected) {
            const rnd = __1.helpers.getRandomInt(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
            yield __1.helpers.sleep(1000 * 60);
            // We don't care about the result we just want to send it every
            // 60 seconds so telegram doesn't stop the connection
            try {
                client._sender.send(new tl_1.Api.Ping({
                    pingId: big_integer_1.default(rnd),
                }));
            }
            catch (e) {
                //await client.disconnect()
            }
            // We need to send some content-related request at least hourly
            // for Telegram to keep delivering updates, otherwise they will
            // just stop even if we're connected. Do so every 30 minutes.
            // TODO Call getDifference instead since it's more relevant
            if (!client._lastRequest ||
                new Date().getTime() - client._lastRequest > 30 * 60 * 1000) {
                try {
                    yield client.invoke(new tl_1.Api.updates.GetState());
                }
                catch (e) { }
            }
        }
    });
}
exports._updateLoop = _updateLoop;
