// telegram/extensions/AsyncQueue.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AsyncQueue = void 0;
class AsyncQueue {
    constructor() {
        this._queue = [];
        this.canPush = true;
        this.resolvePush = (value) => { };
        this.resolveGet = (value) => { };
        this.canGet = new Promise((resolve) => {
            this.resolveGet = resolve;
        });
    }
    push(value) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.canPush;
            this._queue.push(value);
            this.resolveGet(true);
            this.canPush = new Promise((resolve) => {
                this.resolvePush = resolve;
            });
        });
    }
    pop() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.canGet;
            const returned = this._queue.pop();
            this.resolvePush(true);
            this.canGet = new Promise((resolve) => {
                this.resolveGet = resolve;
            });
            return returned;
        });
    }
}
exports.AsyncQueue = AsyncQueue;
