// telegram/extensions/PromisedWebSockets.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromisedWebSockets = void 0;
const websocket_1 = require("websocket");
const async_mutex_1 = require("async-mutex");
const mutex = new async_mutex_1.Mutex();
const closeError = new Error("WebSocket was closed");
class PromisedWebSockets {
    constructor() {
        this.client = undefined;
        this.stream = Buffer.alloc(0);
        this.closed = true;
    }
    readExactly(number) {
        return __awaiter(this, void 0, void 0, function* () {
            let readData = Buffer.alloc(0);
            while (true) {
                const thisTime = yield this.read(number);
                readData = Buffer.concat([readData, thisTime]);
                number = number - thisTime.length;
                if (!number) {
                    return readData;
                }
            }
        });
    }
    read(number) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.closed) {
                throw closeError;
            }
            yield this.canRead;
            if (this.closed) {
                throw closeError;
            }
            const toReturn = this.stream.slice(0, number);
            this.stream = this.stream.slice(number);
            if (this.stream.length === 0) {
                this.canRead = new Promise((resolve) => {
                    this.resolveRead = resolve;
                });
            }
            return toReturn;
        });
    }
    readAll() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.closed || !(yield this.canRead)) {
                throw closeError;
            }
            const toReturn = this.stream;
            this.stream = Buffer.alloc(0);
            this.canRead = new Promise((resolve) => {
                this.resolveRead = resolve;
            });
            return toReturn;
        });
    }
    getWebSocketLink(ip, port) {
        if (port === 443) {
            return `wss://${ip}:${port}/apiws`;
        }
        else {
            return `ws://${ip}:${port}/apiws`;
        }
    }
    connect(port, ip) {
        return __awaiter(this, void 0, void 0, function* () {
            this.stream = Buffer.alloc(0);
            this.canRead = new Promise((resolve) => {
                this.resolveRead = resolve;
            });
            this.closed = false;
            this.website = this.getWebSocketLink(ip, port);
            this.client = new websocket_1.w3cwebsocket(this.website, "binary");
            return new Promise((resolve, reject) => {
                if (this.client) {
                    this.client.onopen = () => {
                        this.receive();
                        resolve(this);
                    };
                    this.client.onerror = (error) => {
                        reject(error);
                    };
                    this.client.onclose = () => {
                        if (this.resolveRead) {
                            this.resolveRead(false);
                        }
                        this.closed = true;
                    };
                    //CONTEST
                    if (typeof window !== "undefined") {
                        window.addEventListener("offline", () => __awaiter(this, void 0, void 0, function* () {
                            yield this.close();
                            if (this.resolveRead) {
                                this.resolveRead(false);
                            }
                        }));
                    }
                }
            });
        });
    }
    write(data) {
        if (this.closed) {
            throw closeError;
        }
        if (this.client) {
            this.client.send(data);
        }
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.client) {
                yield this.client.close();
            }
            this.closed = true;
        });
    }
    receive() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.client) {
                this.client.onmessage = (message) => __awaiter(this, void 0, void 0, function* () {
                    const release = yield mutex.acquire();
                    try {
                        let data;
                        //CONTEST BROWSER
                        data = Buffer.from(yield new Response(message.data).arrayBuffer());
                        this.stream = Buffer.concat([this.stream, data]);
                        if (this.resolveRead) {
                            this.resolveRead(true);
                        }
                    }
                    finally {
                        release();
                    }
                });
            }
        });
    }
}
exports.PromisedWebSockets = PromisedWebSockets;
